Faster 2D shape layer creation

maweigert opened the pull request:
# Description

Currently, the creation of a 2D path or polygon shape layer can be rather slow for large number of shapes (e.g. ~10 s for 5000 polygons, see as well #1562 and #2015).

This PR should improve the shape creation time by a factor ~10-100 (see timing below) and thus allows adding >20k paths/polygons within reasonable time:


https://user-images.githubusercontent.com/11042162/147170582-d7946ad7-50e8-4aec-8205-6b577afe2eaf.mp4


Specifically these parts of the shape layer creation process were changed:

* A) Vectorize/numpify edge mesh generation (`generate_2D_edge_meshes`)
* B) Allow for adding multiple shapes to the dataview at once (`_add_shapes`)
* C) Add support for external triangulation libraries in particular [triangle](https://pypi.org/project/triangle/) (as done by vispy)
* D) Use scikit-image for thumbnail creation (`path_to_mask` and `poly_to_mask`)

### Example

```python

import napari
import numpy as np

def create_sample_coords(n_shapes=5000, n_rays=32):
    radius = 500/np.sqrt(n_shapes)
    center = np.random.uniform(0,1000,(n_shapes,2))
    phi = np.linspace(0,2*np.pi,n_rays)
    rays = np.stack([np.sin(phi), np.cos(phi)],axis=1)
    rays = rays.reshape((1,-1,2))
    rays = rays*np.random.uniform(.9,1.1, (n_shapes,n_rays,2))
    center = center.reshape((-1,1,2))
    coords = center + radius*rays
    return coords

coords = create_sample_coords(5000)

viewer = napari.Viewer()
kwargs = dict(shape_type="path",edge_color="coral")

layer = napari.layers.Shapes(coords,**kwargs)

viewer.add_layer(layer)

```

### Timeing 

`shape_type = 'path'`
|   n_shapes |   main (s) |   PR (s) |   saving % |   speedup  |
|-----------:|---------------------:|--------------------:|-----------:|-----------:|
|       1000 |                1.605 |               0.16  |     90.036 |       10   |
|       3000 |                8.171 |               0.483 |     94.084 |       16.9 |
|      10000 |               65.165 |               1.509 |     97.684 |       43.2 |
|      30000 |              584.644 |               4.674 |     99.201 |      125.1 |

`shape_type = 'polygon'`

|   n_shapes |   main (s) |   PR (s) |   saving % |   speedup  |
|-----------:|------------------------:|-----------------------:|-----------:|-----------:|
|       1000 |                   5.822 |                  0.269 |     95.38  |       21.6 |
|       3000 |                  22.162 |                  0.811 |     96.341 |       27.3 |
|      10000 |                 132.089 |                  2.647 |     97.996 |       49.9 |
|      30000 |                1050.56  |                  9.479 |     99.098 |      110.8 |



## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

maweigert committed:
Vectorize edge mesh generation in shape layer

maweigert committed:
Allow adding multiple shapes at once to dataview (faster)

maweigert committed:
Vectorize some shapes code

maweigert committed:
Speed up text coordinates when no text element present

maweigert committed:
Improve path_to_mask speed

maweigert committed:
Add external triangulation library in shapes layer

maweigert committed:
Use skimage for thumbnail rendering (introduce skimage dep)

maweigert committed:
Fix center bug, another vectorization

andy-sweet commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
+            # same but faster than:
+            # index = np.repeat([[shape_index, 0]], len(triangles), axis=0)
+            index = np.zeros((len(triangles), 2), np.int32)
+            index[:, 0] = shape_index
```
Nit: this 4-line block is repeated four times and re-reading the same implementation comment makes the interesting stuff harder to parse/understand.

You solve this by defining a helper function like

```python
def _make_index(shape_index, length):
    """Same but faster than `np.repeat([[shape_index, 0]], length, axis=0)`"""
    index = np.zeros((length, 2), np.int32)
    index[:, 0] = shape_index
    return index
```

and calling it instead.

Another potential advantage is that we call this in `add` for similar speed ups.

maweigert replied:
Yes, makes sense. I adjusted the code

andy-sweet commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
+        )
+
+        if z_refresh:
+            # Set z_order
```
I know this is also in `add`, but this implementation comment doesn't add anything here - the code is readable on its own.

andy-sweet commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
+            all_index.append(index)
+
+            # Add faces to mesh
+            m = m_mesh_vertices_count
```
Optional: consider a more meaningful variable name than `m` because in order to understand what this does/means, I have to jump ahead about 10 lines of code.

I also wonder how much we really need both `m` and `m_mesh_vertices_count` and if instead we can rely on the lengths of some of the arrays here instead.

maweigert replied:
Agreed. I simply reproduced the existing code of `add()` in a vectorized form, without trying to hard to understand what is really going on ;) So for someone actually knowing what is going on, there might be obvious tweaks.

andy-sweet commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
+            vertices = shape._face_vertices
+            all_mesh_vertices.append(vertices)
+            m_mesh_vertices_count += len(vertices)
+            vertices = shape._face_vertices
```
Optional: I realize this is from `add`, but consider removing many of the assignments to `vertices` because it's hard to reason about whether this actually affects code after this block (I think it doesn't).

In this case, you could just write `all_mesh_vertices_centers.append(shape._face_vertices)`, which seems fine to me. I think there are other cases below that could also be rewritten this way without an effect on behavior.

andy-sweet commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -533,7 +534,23 @@ def triangulate_face(data):
     """
-    vertices, triangles = PolygonData(vertices=data).triangulate()
+
+    try:
```
Is it possible to concisely summarize the difference between this and [vispy's `Triangulation.triangulate`](https://cs.github.com/vispy/vispy/blob/26256fdc2574259dd227022fbce0767cae4e244b/vispy/geometry/triangulation.py?q=import%20triangle#L822)? And why we don't use that instead?

At the least, I think it would be useful to have some link to the vispy code and an explanation of what is different here.

andy-sweet reviewed (commented):
Overall, the impact of this PR is fantastic and the numbers speak for themselves. I pulled down the branch and tried it with and without `triangle` and didn't notice huge changes for small numbers (i.e. <10) of polygons, but I guess that's roughly expected (i.e. the performance benefits become much more noticeable for large numbers of polygons)

If you want to confirm whether this fixes #1058, you can use the following as the beginning of a minimal reproducing test:

```python
def test_bad_poly():
    data = [
        np.array(
            [
                [4, 4],
                [3, 4],
                [1, 4],
                [4, 4],
                [1, 2],
            ]
        )
    ]
    Shapes(data, shape_type='polygon')
```

I'm not sure what to assert yet, but see [the vispy issue](https://github.com/vispy/vispy/issues/2247) for a diagram of the geometry described.

Most of my line comments are also related to existing code in `ShapeList.add`, but I still think we should fix those things even if it requires refactoring/changing `add` a little.

kevinyamauchi commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
+        edge_colors: iterable of edge_color
+        z_refresh : bool
+            If set to true, the mesh elements are reindexed with the new z order.
+            When shape_index is provided, z_refresh will be overwritten to false,
+            as the z indices will not change.
+            When adding a batch of shapes, set to false  and then call
+            ShapesList._update_z_order() once at the end.
```
It looks like this might be leftover from the `add()` method.

maweigert replied:
Should be the adjusted docstring for multiple shapes (similar but not identical to `add()` )

kevinyamauchi commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
+
+            if edge_color is None:
+                edge_color = np.array([0, 0, 0, 1])
+            all_edge_color.append(edge_color)
```
Minor optimization, but this feels a bit redundant, as `face_color` and `edge_color` are just unpacked from `face_colors` and `edge_colors` above.

Did you have a use case in mind `None` or is the left over from the `add()` method? If it was just to give a default value, perhaps that could be handled before the loop (i.e., create an array of the default colors that matches the number of shapes). However, since the `ShapeList` is a private attribute of the `Shapes` layer (i.e., the user isn't directly interacting with it), I would be okay with requiring the face and edge colors to be passed in.

maweigert replied:
Yes, that's a good idea. I added a default value array before the loop

kevinyamauchi commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
+    def add_multiple(
+        self,
+        shapes,
+        face_colors=None,
+        edge_colors=None,
```
These default values seem like they would cause an error in the for loop on line 399.

maweigert replied:
Correct, thanks for catching! (should be fixed now)

kevinyamauchi commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +353,178 @@ def add(
             # Set z_order
             self._update_z_order()
 
+    def add_multiple(
```
This method is getting pretty long. It might be nice to break it up into smaller util/helper functions and then `add_multiple()` can dispatch them. It would likely make this easier to read and also easier to share functions between `add()` and `add_multiple()`.

I understand this is a lot of carry over from the current implementation of the `add()` method. However, this seems like a good opportunity for a refactor.

kevinyamauchi replied:
Along similar lines, do you think it makes sense to either get rid of `add()` or make `add()` a passthrough to `add_multiple()` (with any required coercion, etc.)?

maweigert replied:
(Reactivating this PR, sorry for the delay)

Yes, ideally there should be only one method adding shapes. Currently, `add()` allows for adding shapes at a specific index, whereas `add_multiple()` will add consecutive indices at the end. It might be easy to merge those two, but I had (and still don't have) know real idea of what the purpose of the `shape_index` parameter has been and what side effects to expect, which is why I left `add()` as original.

sofroniewn replied:
I think at the time `shape_index` seemed like a reasonable thing we might want - but if we ended up not using it then we could refactor it out.

kevinyamauchi reviewed (commented):
Thank you @maweigert - this is great stuff! The benchmarks are pretty enticing. Overall, I like the direction this is going. I left some high level comments below. I saw you just marked this PR as WIP, so please feel free to disregard if they are outdated. I will be offline until the new year, but I am happy to chat about this when I get back. Happy holidays!

Carreau committed:
minor optimisations

maweigert committed:
merge from upstream and add shape creation demo

Carreau committed:
add test for generate_3D_edge_meshes

Also add example script to visualized triangulated edges with various
parameters to visually check any changes that could be introduced.

In particular we see that the current method path do not properly
overlap with some combinations. That might need to be fixed later, but
will mostly be used for working on #3867.

minor refactor

maweigert committed:
Fix issue for beveled and closed paths

maweigert committed:
Merge

maweigert committed:
Small change

Carreau committed:
update expected

Carreau committed:
add function to regenerate test cases.

Carreau committed:
Bump scikit image version to 0.19.1,

Tests requires =! 0.19.0, so the next avail version is 0.19.1 (dec 15th)

Carreau commented on examples/dev/plot_2d_edge_meshes.py:
```diff
@@ -0,0 +1,36 @@
+
+fig, axes = plt.subplots(2, 3)
+# fig.set_figwidth(15)
+# fig.set_figheight(10)
```
```suggestion
```

maweigert committed:
Small change in demo script

maweigert committed:
Allow None colors in shape list

maweigert committed:
Small tweaks

maweigert committed:
Faster update_colors

Carreau commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -571,6 +748,19 @@ def update_face_color(self, index, face_color, update=True):
+        self._mesh.triangles_colors[all_indices] = self._face_color[
+            self._mesh.triangles_index[all_indices, 0]
+        ]
+        if update:
+            self._update_displayed()
```
Note for later (not this pr),
I see this pattern used in many places, I _think_ it would be better to turn it into a reentrant context manager `self.batch_update()` that execute all the `_update_display`/ `_update_z_order` on exit.

Carreau commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,171 @@ def add(
+            Each Shape must be a subclass of Shape, one of "{'Line', 'Rectangle',
+            'Ellipse', 'Path', 'Polygon'}"
+        face_colors: iterable of face_color
+        edge_colors: iterable of edge_color
```
```suggestion
        face_colors : iterable of face_color
        edge_colors : iterable of edge_color
```

Carreau commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,171 @@ def add(
             # Set z_order
             self._update_z_order()
 
+    def add_multiple(
```
I believe this can be used arround line 98:

```
    @ndisplay.setter
    def ndisplay(self, ndisplay):
        if self.ndisplay == ndisplay:
            return

        self._ndisplay = ndisplay
        self._mesh.ndisplay = self.ndisplay
        self._vertices = np.empty((0, self.ndisplay))
        self._index = np.empty((0), dtype=int)
        for index in range(len(self.shapes)):
            shape = self.shapes[index]
            shape.ndisplay = self.ndisplay
            self.remove(index, renumber=False)
            self.add(shape, shape_index=index)
        self._update_z_order()

```

This looks like we can just flush all shapes and call add_multiple

Line 764

```
    def update_dims_order(self, dims_order):
        """Updates dimensions order for all shapes.

        Parameters
        ----------
        dims_order : (D,) list
            Order that the dimensions are rendered in.
        """
        for index in range(len(self.shapes)):
            if not self.shapes[index].dims_order == dims_order:
                shape = self.shapes[index]
                shape.dims_order = dims_order
                self.remove(index, renumber=False)
                self.add(shape, shape_index=index)
        self._update_z_order()

```

We can maybe also just filter all the shaped and set them all instead of adding/removing individually.

the other few places that are using shape_index, are calling self.remove then self.add at same index to transform/scale a shape. 

I think in those case it's ok to not re-add at the same index unless we want the shape/index mapping to always stay the same.

maweigert replied:
> I't weird as it appear I can't push on PRs anymore (neither this one not others), but I can merge PRs. Was there any change of permissions recently ?

Yes, that's weird. @jni @sofroniewn anyone knows the reason?

sofroniewn replied:
I don't - I don't think we changed anything on our end, maybe a GitHub update. @jni?

jni replied:
I see the Quansight folks including @Carreau as having write access, which I *think* is what it should take...?

At any rate, @Carreau @maweigert there's some merge conflicts from #4416, and I don't know a priori how to merge them correctly, would be good if you could merge in main here. 0.4.16rc0 is in ~36h!

jni replied:
I'll be around this evening if anyone wants me to push anything anywhere 😂

maweigert replied:
@Carreau seem like you are able to push now?

Carreau replied:
Yep, we sorted it out with JNI an hour ago.

Carreau commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,171 @@ def add(
+        m_mesh_vertices_count = len(self._mesh.vertices)
+
+        if face_colors is None:
+            face_colors = np.tile(np.array([1, 1, 1, 1]), (len(shapes), 1))
```
I think that will have ~0 impact on perf, but `np.ones((len(shapes), 1))` will likely be faster/more readable.

maweigert replied:
`np.ones((len(shapes), 4))`?

Carreau replied:
sure.

maweigert replied:
As we need `np.tile` for `edge_colors` I kept is for consistency

Carreau commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,171 @@ def add(
+                        'shape must be subclass of Shape',
+                        deferred=True,
+                    )
+                )
```
note to self, maybe move error handling outside the for loop to not add only part of the shapes internally

maweigert replied:
Yes, makes sense. Done

Carreau commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,171 @@ def add(
+            all_z_index.append(shape.z_index)
+
+            all_face_color.append(face_color)
+            all_edge_color.append(edge_color)
```
Those two seem unneeded they will end up being identical to `face_colors` and `edge_colors` therefor .... (next line 484)

maweigert replied:
Yes, fixed

Carreau commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,171 @@ def add(
+        )
+        self._edge_color = np.vstack(
+            (self._edge_color, np.vstack(all_edge_color))
+        )
```
(from 428)... here you can just use `face_colors`, and `edge_colors`.

maweigert replied:
See above

maweigert committed:
Update napari/layers/shapes/_shape_list.py

Co-authored-by: Matthias Bussonnier <bussonniermatthias@gmail.com>

Carreau committed:
Merge remote-tracking branch 'origin/main' into feat-faster-shapes

Carreau committed:
emty test

Carreau reviewed (commented):
FOrgot to send those comments last week. I'll try to take care of a few of them.

maweigert committed:
Fix test for empty text

maweigert committed:
Add ellipse shape type to shape demo

maweigert committed:
Add ellipse shape type to shape demo

maweigert committed:
Test for shapeness outside of loop

Carreau committed:
Remove some intermediate variable assignment.

it reuses the same variable name many time which make it confusing to
read.

Carreau commented on examples/dev/demo_shape_creation.py:
```diff
@@ -0,0 +1,110 @@
+
+    kwargs = dict(
+        shape_type=args.type,
+        properties=properties if args.properties else None,
```
This seem to be problematic now, it does not work with features. Investigating.

Carreau replied:
sorry, does not work when using concat.... I think we can just remove concat.

maweigert replied:
I changed the concat option

Carreau replied:
+1 also had some issues as my editor kept a old version that was autosaved. Looks good. I just pushed minor changes in intermediate variable in the main function to make it easier to read.

maweigert committed:
Change shape demo

Carreau committed:
Merge remote-tracking branch 'maweigert/feat-faster-shapes' into feat-faster-shapes

sofroniewn reviewed (approved):
This is great work @maweigert @Carreau! Very excited to get this in. We can keep improving things in shapes as they come up if we need too. 🚀

jni commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,170 @@ def add(
+        `shape_index`, whereas `add_multiple` will append them as a full batch
+        """
+
+        def _make_index(length, shape_index, cval=0):
+            """Same but faster than `np.repeat([[shape_index, cval]], length, axis=0)`"""
+            index = np.empty((length, 2), np.int32)
+            index.fill(cval)
+            index[:, 0] = shape_index
+            return index
```
Why is this function nested?

maweigert replied:
since it's an arcane idiosyncratic helper function specific to the enclosing function :)

jni replied:
ie no reason :joy: We should pull it out. One of the NASA programming principles for bug-free code is functions should be 25 lines or less. Nested functions make that ~impossible to achieve. Obviously there's other bits that need compressing in this function but this is a trivial win.

maweigert replied:
I would agree in general. But the semantic scope of `make_index` is clearly confined to the enclosing function (i.e. useless to the rest of the code) and thus pulling it out would imho actually decrease maintainability

jni commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,170 @@ def add(
+        all_mesh_vertices_index = list()
+        all_mesh_triangles = list()
+        all_mesh_triangles_index = list()
+        all_mesh_triangles_colors = list()
```
```suggestion
        all_z_index = []
        all_vertices = []
        all_index = []
        all_mesh_vertices = []
        all_mesh_vertices_centers = []
        all_mesh_vertices_offsets = []
        all_mesh_vertices_index = []
        all_mesh_triangles = []
        all_mesh_triangles_index = []
        all_mesh_triangles_colors = []
```

This is both more Pythonic and [faster](https://www.thecodeteacher.com/question/2324/python---Why-is-[]-faster-than-list())!

maweigert replied:
True, don't even know why is was like this in the first place ;)

jni commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,170 @@ def add(
+        else:
+            face_colors = np.asarray(face_colors)
+
+        if edge_colors is None:
+            edge_colors = np.tile(np.array([0, 0, 0, 1]), (len(shapes), 1))
+        else:
+            edge_colors = np.asarray(edge_colors)
```
Just checking my understanding: if we are colouring by properties, currently, we still need a "backup" direct colour array to be present, and that's where this will go?

jni commented on napari/layers/shapes/_shape_list.py:
```diff
@@ -353,6 +352,170 @@ def add(
+                )
+            )
+
+        for shape, face_color, edge_color in zip(
+            shapes, face_colors, edge_colors
+        ):
```
I have to agree with @kevinyamauchi, this function is so long. I really have trouble following it. It would be great if it was refactored into some helper functions.

maweigert replied:
Yeah, I agree. It's still like this as I only "translated" the existing `add()` function into a version accepting multiple shapes, without attempting to refactor the details in fear of possible side effects

jni commented on setup.cfg:
```diff
@@ -111,7 +112,6 @@ testing =
     fsspec
     zarr
     matplotlib
-    scikit-image>=0.18.1,!=0.19.0
```
Probably there needs to be `triangle` added to a section... Maybe a new section [optional] or [performance]?

maweigert replied:
we could do that, if as an *optional* dep we circumvent the [license issue above](https://github.com/napari/napari/pull/3867#issuecomment-999978209)? @sofroniewn

sofroniewn replied:
i believe these are just "testing" dependencies and so as we don't ship with them we don't have to worry about the license issue here

maweigert replied:
Yes, but its not really related to `testing` per se (btw, `pip install[testing]` seems to fail on M1 anyhow to `tensorstore` not being able to be built). Maybe another section (as I understood @jni) makes sense indeed?

sofroniewn replied:
yup yup, sorry - i guess maybe doing `[performance]` is fine. i think the optional aspect of it makes it ok for licensing. if we make it my default in the bundle then we might need to add something special, but that can come later

maweigert replied:
Would `all` be too much?

sofroniewn replied:
`all` means we'll hit that license issue for sure - not sure if we're ready to commit to that

maweigert replied:
Yeah, I agree. Maybe it's best to avoid specifying `triangle` as a dependency at all and let users simply install if for themselves? ([this seems to be the way vispy does it](https://vispy.org/installation.html#optional-dependencies))

sofroniewn replied:
yeah, let's just hold off for now - we can do something for next release

jni replied:
@sofroniewn @maweigert PySide2 is LGPL2 anyway so I don't understand what we are worried about here. And note that "all" currently depends on PyQt5 which is GPL!

maweigert replied:
@sofroniewn @jni Ok, I feel like getting opposing gradients here ;) 

I think immediate options for now are

A) Don't add any dep and let user `pip install triangle`
B) Add it to `[all]`, thus effectively making the dep opt-out  (as most people will probably do `pip install napari[all]` anyway)
C) Create a separate dep section that inherits from `[all]` while adding even more opt-in deps such as `triangle` (might be thus called `[uberall]`)

jni replied:
I think I like (D) add [performance]. We can later include it or not include it in [all], but for now it makes it easier to say `pip install napari[all,performance]`. (I do think it should go in [all] for the record, but that will slow down discussion.)

And yes you're getting opposing gradients but imho mine is the correct gradient, @sofroniewn's is just noise. :stuck_out_tongue_winking_eye: :joy: More seriously, I really don't see how this in any way affects our licensing, since we already depend on an LGPL package even for the bundle, and a GPL package in [all].

maweigert replied:
Hahah, ok :) 

So I opted for (D'), which is your suggestion of adding a new section, but I named it `[optional]` (which might make it more transparent).

jni reviewed (approved):
Sorry, I promised @sofroniewn that I would have a pass at making the changes I'd like to see here myself, but didn't get time. I've noted a few proposed changes. I'll also say that my preference would be for add() to do double duty, which could be achieved with the following refactor:

- rename add to _add_classic_TM 😜 
- rename add_multiple to _add_multiple
- write a new add() function that takes in the union of the arguments of add and add_multiple. Then if given more than one shape, call _add_multiple, if given one shape, wrap it in a list and call _add_multiple, if given one shape and a shape_index, call _add_classic_TM, and if given more than one shape and also a shape_index, raise.

Personally I think I'd be ok with missing out another release cycle in order to get this right, but if others really want to push it I'm ok with merging despite my review and attempting to fix soon.

maweigert committed:
Update napari/layers/shapes/_shape_list.py

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

maweigert committed:
Add single entry point for single and multiple shape add

maweigert committed:
Remove type hint (debugging)

maweigert committed:
Add optional dependency section

maweigert committed:
Readd typing
